week2.6.5.group_by.py

```diff
@@ -0,0 +1,19 @@
+# region group by - this 2
+
+def group_by(func, iterable):
+    """
+    Returns a dictionary where the keys are the result of applying the input function
+    on each item in the input iterable, and the values are a list of items from the
+    input iterable that result in the same key.
+    """
+    result = {}
+    for item in iterable:
+        key = func(item)  # Apply the input function on the item to get the key
+        if key not in result:
+            result[key] = [item]  # If key not in dictionary, add a new key-value pair
+        else:
+            result[key].append(item)  # If key already in dictionary, append item to existing value
+    return result

```

Comment:
You should aim to write everything in a single line, even if you can. 
This link https://towardsdatascience.com/you-should-stop-craving-for-python-one-liners-5dbf8e0de23d explains it well, even though I disagree with how important memory and runtime are in most python programs.